crates/sema/src/typeck/mod.rs

```diff
@@ -11,13 +12,91 @@ pub(crate) fn check(gcx: Gcx<'_>) {
         gcx.sess,
         gcx.hir.par_contract_ids().for_each(|id| {
             check_duplicate_definitions(gcx, &gcx.symbol_resolver.contract_scopes[id]);
+            check_storage_size_upper_bound(gcx, id);
         }),
         gcx.hir.par_source_ids().for_each(|id| {
             check_duplicate_definitions(gcx, &gcx.symbol_resolver.source_scopes[id]);
         }),
     );
 }
 
+/// Checks for violation of maximum storage size to ensure slot allocation algorithms works.
+/// Reference: https://github.com/ethereum/solidity/blob/03e2739809769ae0c8d236a883aadc900da60536/libsolidity/analysis/ContractLevelChecker.cpp#L556C1-L570C2
+fn check_storage_size_upper_bound(gcx: Gcx<'_>, contract_id: hir::ContractId) {
+    let contract_span = gcx.hir.contract(contract_id).span;
+    let contract_items = gcx.hir.contract_items(contract_id);
+    let mut total_size = U256::ZERO;
+    for item in contract_items {
+        if let hir::Item::Variable(variable) = item {
+            // Skip constant and immutable variables
+            if variable.mutability.is_none() {
+                let t = gcx.type_of_hir_ty(&variable.ty);
+                match ty_upper_bound_storage_var_size(t, gcx)
+                    .and_then(|size_contribution| total_size.checked_add(size_contribution))
+                {
+                    Some(sz) => {
+                        total_size = sz;
+                    }
+                    None => {
+                        gcx.dcx()
+                            .err("contract requires too much storage")
+                            .span(contract_span)
+                            .emit();
+                        return;
+                    }
+                }
+            }
+        }
+    }
+
+    if gcx.sess.storage_sz_ub {
```

Comment:
let's call it like "print-contract-storage-sizes"

Reply:
We only check the upper bound of storage size, so I'm afraid that wouldn't be accurate depiction of what users are looking for. 

for example:
If there's 

```
int128 a;
int128 b;
```
Currently, this totals to "2", which is less than 2^256 so passes the check.

Whereas some wanting to print storage size would expect 128 + 128 = 256 bits so `1`

If we go with this, do we make another flag for the _actual_ contract storage size then?

Reply:
Tbh, I feel like this check isn't necessary at all lol ! I don't why solidity does it... Maybe I should convert it to actual contract storage size haha

Reply:
@DaniPopes isn't this more like "print-contract-max-storage-sizes" ?